customer-service/src/main/java/io/eventuate/examples/tram/ordersandcustomers/customers/web/OrderEventHttpConsumer.java

```diff
@@ -0,0 +1,33 @@
+package io.eventuate.examples.tram.ordersandcustomers.customers.web;
+
+import io.eventuate.examples.tram.ordersandcustomers.customers.service.CustomerService;
+import io.eventuate.examples.tram.ordersandcustomers.orders.domain.events.OrderCreatedEvent;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RestController;
+import reactor.core.publisher.Mono;
+
+@RestController
+public class OrderEventHttpConsumer {
+
+  private CustomerService customerService;
+
+  @Autowired
+  public OrderEventHttpConsumer(CustomerService customerService) {
+    this.customerService = customerService;
+  }
+
+  @RequestMapping(value = "/events/orderserviceevents/io.eventuate.examples.tram.ordersandcustomers.orders.domain.Order/{aggregateId}/io.eventuate.examples.tram.ordersandcustomers.orders.domain.events.OrderCreatedEvent/{eventId}", method = RequestMethod.POST)
+  public Mono<Void> handleOrderCreatedEvent(@RequestBody OrderCreatedEvent event, @PathVariable("aggregateId") String aggregateId) {
```

Comment:
What support is there for handling duplicate events?
e.g. what would happen if the OrderCreatedEvent was delivered twice?

BTW If the primary key of a CreditReservation was (customerId, orderId) this might help

Reply:
Chris, there is duplicate publishing detection: https://github.com/eventuate-tram/eventuate-tram-messaging-http/blob/master/eventuate-tram-http-spring-consumer/src/main/java/io/eventuate/tram/http/spring/consumer/duplicatedetection/IdempotentHandlerAspect.java

Reply:
Is it used in the `Customer Service`?
If so what about the reactive vs. non reactive transaction issue?
This also means that the event handlers should not manage their own transactions.

Reply:
Is it used in the `Customer Service`?
If so what about the reactive vs. non reactive transaction issue?
This also means that the event handlers should not manage their own transactions.

Reply:
Chris, I do not understand. Http proxy delivers events to customer service, if event is duplicated it is not delivered to customer service.

Reply:
The HTTP proxy cannot do duplicate detection. Only the (Customer) Service can determine whether an event is a duplicate.

Consider the following scenario:

1. Proxy does HTTP POST to Customer Service
2. Customer service successfully updates database etc.
3. Customer Service sends back an HTTP response
4. Proxy crashes before receiving the HTTP response and recording the duplicate.
5. Proxy recovers and redelivers what is now a duplicate.

Reply:
Thank you for clarification, now I understand.

> BTW If the primary key of a CreditReservation was (customerId, orderId) this might help

In JPA composite keys are implemented as embedded objects (at least, I remember that is necessary to create a class that contains key fields), but r2dc does not support embedding. I will check, probably composite pk handled in other way.

Reply:
No, it is not supported:

https://gitmemory.com/issue/spring-projects/spring-data-r2dbc/158/517294119
https://github.com/spring-projects/spring-data-jdbc/issues/574

Reply:
> Thank you for clarification, now I understand.
> 
> >

Please file an issue to remove pointless duplicate key detection from the HTTP proxy.

Reply:
> but r2dc does not support embedding

Spring Data R2DBC repositories might not support composite primary keys but r2dbc is a reactive library for executing SQL statements. The SQL statements can do what ever you want. This is some very simple database access code for loading/saving CreditReservations.

Reply:
So, I need to keep spring data for customer, but replace it by sql for reservations, right?

Reply:
> So, I need to keep spring data for customer, but replace it by sql for reservations, right?

What are the options and their respective trade-offs. Which one do you think is best?

Reply:
Other option is to completely remove spring data. it has only one advantage: more consistent approach. However, I think it is ok to use simple high-level approach where it is possible and lower-level sql where it is not.

Reply:
> Please file an issue to remove pointless duplicate key detection from the HTTP proxy.
Done: https://github.com/eventuate-tram/eventuate-tram-messaging-http/issues/13